hdinsight/hdinsight.go

```diff
@@ -0,0 +1,128 @@
+	}
+	client := hdinsight.NewClustersClient(config.SubscriptionID())
+	client.Authorizer = a
+	client.AddToUserAgent(config.UserAgent())
+	return &client, nil
+}
+
+// StorageAccountInfo describes the storage account used for the cluster's file system.
+type StorageAccountInfo struct {
+	Name      string
+	Container string
+	Key       string
+}
+
+// CreateHadoopCluster creats a simple hadoop 3.6 cluster
+func CreateHadoopCluster(resourceGroup, clusterName string, info StorageAccountInfo) (*hdinsight.Cluster, error) {
+	client, err := getClustersClient()
+	if err != nil {
+		return nil, err
+	}
+	// the default duration is 15 minutes which is just a tad too short
+	client.PollingDuration = 20 * time.Minute
+	util.PrintAndLog("creating hadoop cluster")
+	future, err := client.Create(context.Background(), resourceGroup, clusterName, hdinsight.ClusterCreateParametersExtended{
+		Location: to.StringPtr(config.Location()),
+		Properties: &hdinsight.ClusterCreateProperties{
+			ClusterVersion: to.StringPtr("3.6"),
+			OsType:         hdinsight.Linux,
+			Tier:           hdinsight.Standard,
+			ClusterDefinition: &hdinsight.ClusterDefinition{
+				Kind: to.StringPtr("hadoop"),
+				Configurations: map[string]map[string]interface{}{
+					"gateway": {
+						"restAuthCredential.isEnabled": true,
+						"restAuthCredential.username":  "admin",
+						"restAuthCredential.password":  "Thisisalamepasswordthatwillberemoved2.",
+					},
+				},
+			},
+			ComputeProfile: &hdinsight.ComputeProfile{
+				Roles: &[]hdinsight.Role{
+					hdinsight.Role{
+						Name:                to.StringPtr("headnode"),
+						TargetInstanceCount: to.Int32Ptr(2),
+						HardwareProfile: &hdinsight.HardwareProfile{
+							VMSize: to.StringPtr("Large"),
+						},
+						OsProfile: &hdinsight.OsProfile{
+							LinuxOperatingSystemProfile: &hdinsight.LinuxOperatingSystemProfile{
+								Username: to.StringPtr("clusteruser"),
+								Password: to.StringPtr("Thisisalamepasswordthatwillberemoved1."),
+							},
+						},
+					},
+					hdinsight.Role{
+						Name:                to.StringPtr("workernode"),
+						TargetInstanceCount: to.Int32Ptr(1),
+						HardwareProfile: &hdinsight.HardwareProfile{
+							VMSize: to.StringPtr("Large"),
+						},
+						OsProfile: &hdinsight.OsProfile{
+							LinuxOperatingSystemProfile: &hdinsight.LinuxOperatingSystemProfile{
+								Username: to.StringPtr("clusteruser"),
+								Password: to.StringPtr("Thisisalamepasswordthatwillberemoved1."),
+							},
+						},
+					},
+					hdinsight.Role{
+						Name:                to.StringPtr("zookeepernode"),
+						TargetInstanceCount: to.Int32Ptr(3),
+						HardwareProfile: &hdinsight.HardwareProfile{
+							VMSize: to.StringPtr("Small"),
+						},
+						OsProfile: &hdinsight.OsProfile{
+							LinuxOperatingSystemProfile: &hdinsight.LinuxOperatingSystemProfile{
+								Username: to.StringPtr("clusteruser"),
+								Password: to.StringPtr("Thisisalamepasswordthatwillberemoved1."),
```

Comment:
Server side validation.  It could be the case that the applicable size will depend on the type of cluster being created.  Hard to say and the docs for this are very light on the details. :(